File: Matlab/DeepGLM/nnfun/nnActivationGrad.m
function out = nnActivationGrad(z,func)
%NNACTIVATIONGRAD Calculate derivative of activation output at hidden nodes 
%in each backward pass
%
%   Copyright 2018 Minh-Ngoc Tran (minh-ngoc.tran@sydney.edu.au) and Nghia
%   Nguyen (nghia.nguyen@sydney.edu.au)
%   
%   http://www.xxx.com
%
%   Version: 1.0
%   LAST UPDATE: April, 2018


switch func
    case 'Linear'
        out = ones(size(z));
    case 'Sigmoid'
        temp = activation(z,text);
        out = temp.*(1-temp);
    case 'Tanh'
        temp = activation(z,text);
        out = 1 - temp^2;
    case 'ReLU'
        out = z>0;
    case 'LeakyReLU'
        if z > 0
            out = 1;
        else 
            out = alpha;
        end
end

end

File: R/02_libs/nnActivationGrad.R
# Function to calculate derivative of activation function
# Input must be a matrix
nnActivationGrad <- function(a,func){
  switch (func,
          Linear = {out <- matrix(1,nrow(a),ncol(a))},
          Sigmoid = {out <- 1/(1+exp(-a))},
          ReLU = {out <- (a>0)*1},
          defaut)
  return(out)
}
